tsdb/engine/tsm1/engine.go

```diff
@@ -223,50 +228,89 @@ func (e *DevEngine) Begin(writable bool) (tsdb.Tx, error) {
 
 func (e *DevEngine) WriteTo(w io.Writer) (n int64, err error) { panic("not implemented") }
 
-func (e *DevEngine) compact() {
-	for {
+func (e *DevEngine) writeSnapshot() error {
+	// Lock and grab the cache snapshot along with all the closed WAL
+	// filenames associated with the snapshot
+	closedFiles, snapshot, err := func() ([]string, *Cache, error) {
+		e.mu.Lock()
+		defer e.mu.Unlock()
+
 		if err := e.WAL.CloseSegment(); err != nil {
-			e.logger.Printf("error rolling current WAL segment: %v", err)
-			time.Sleep(time.Second)
-			continue
+			return nil, nil, err
 		}
 
-		tsmFiles, segments, err := e.CompactionPlan.Plan()
+		segments, err := e.WAL.ClosedSegments()
 		if err != nil {
-			e.logger.Printf("error calculating compaction plan: %v", err)
-			time.Sleep(time.Second)
-			continue
+			return nil, nil, err
 		}
 
-		if len(tsmFiles) == 0 && len(segments) == 0 {
+		snapshot := e.Cache.Snapshot()
+
+		return segments, snapshot, nil
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	go func() {
```

Comment:
Yes.  Needs a read lock because it's accessing fields on the engine.  We'll trigger races if a compaction is running while engine is closed (e.g. shard/db dropped).

I think you can make `Compactor` an local instance an remove it from the engine fields.  Then just take the lock before calling `FileStore.Replace`.